File: cpp/1291_SequentialDigits.cpp
class Solution {
public:
    vector<int> sequentialDigits(int low, int high) {
        vector<int> result;
        if (high <= low) 
            return result;
        
        int divisor = 10, add = 1;
        int begin = 1;
        while (low / divisor > 0) {
            divisor *= 10;
            begin = (begin * 10) + (begin % 10 + 1);
            add = (add * 10) + 1;
        }
        
        int start = begin;
        while (start <= high) {
            if (start >= low)
                result.push_back(start);

            start += add;
            if (start % 10 == 0) {
                begin = (begin * 10) + (begin % 10 + 1);
                start = begin;
                add = add * 10 + 1;
            }
        }
        return result;
    }
};

File: swift/1291_SequentialDigits.swift
class Solution {
    func sequentialDigits(_ low: Int, _ high: Int) -> [Int] {
        var result: [Int] = [Int]()
        if high <= low { return result }
        
        var divisor: Int = 10, additor = 1, start = 1;
        
        // beginning of sequence for n digits of low
        while low / divisor > 0 {
            start = (start * 10) + (start % 10 + 1)
            divisor *= 10
            additor = additor * 10 + 1
        }
        
        var begin: Int = start
        
        // keep adding to begin until high
        while begin <= high {
            if begin >= low { result.append(begin) }
            begin += additor
            if begin % 10 == 0 {
                start = (start * 10) + (start % 10 + 1)
                begin = start
                additor = additor * 10 + 1
            }
        }
        return result
    }
}


